### ISSUE TITLE
[PubSub] Sometimes subscriber.stop.wait! hangs forever

### ISSUE THREAD

**Issue by hellupline** [GoogleCloudPlatform/google-cloud-ruby#2206]
Using the example code below,  `subscriber.stop.wati!` hangs waiting for some thread to stop, blocking the graceful shutdown of the subscriber

Example code: 
```ruby
subscription = topic.subscription("demo")
subscriber = subscription.listen do |msg|
  puts "begin #{msg.data}"
  sleep 5  # fake msg processing
  puts "end #{msg.data}"
end

subscriber.start
sleep 60.seconds
puts "time to finish"
subscriber.stop.wait!
```
this example code runs the subscriber for 60 seconds, then try to do a graceful shutdown, it always gets to the `stop.wait!`, sometimes, it does finish the script, sometimes it hangs indefinitely

this behaviour happens with both emulator and cloud-pubsub

**Comment by quartzmo** [comment]
@hellupline Thank you for reporting this issue!  Is there more you can tell us about your system, including Ruby and google-cloud-pubsub versions?

**Comment by hellupline** [comment]
sorry for the delay @quartzmo , 

ruby version: 2.4.3
google-cloud-pubsub gem version: 0.29.0
google-cloud gem version: 0.49.0

environmnet: kubernetes engine at google cloud platform, high cpu instance type
region us-east4

**Comment by blowmage** [comment]
@hellupline You had 100% errors on the StreamingPull API? What types of errors?

**Comment by blowmage** [comment]
@hellupline @vnbrs how reliable is the subscriber hanging for you? Does it happen every time? Usually? Occasionally? Rarely?

**Comment by blowmage** [comment]
I'm still looking into this. I have not been able to fully reproduce it yet, but I have some ideas on possible causes. Any additional information you can share would be very, very helpful.

**Comment by hellupline** [comment]
sorry for the delay, 

I after some struggle, we decide to update all gems related, and it kind of fixed both issues, where  some threads stopped and google pubsub api errors where too high

but before the upgrade,our GCP console was 100% error rate ( I dont remember the error code, sorry ), but the lib did not report any error, even with the debug logging,

the erorrs reports on GPC console skyrocketed at same instant I started the script, so I assume it was related

the faulty version was 0.29, the currently is 0.31, so It was fixed between those 

thankyou for you atention tho

I think we can close this issue

**Comment by blowmage** [comment]
@vnbrs What version of the gem are you running?

**Comment by vnbrs** [comment]
I was running with `0.29` and I've updated it to `0.31`, along with `grpc` to `1.8.7`. Problem solved since then.

**Comment by blowmage** [comment]
Okay, there is an issue where an older GRPC library can fail for internal reasons, and we started recovering from that in google-cloud-pubsub 0.30.2. I had thought it was in an earlier release, but I just verified it was definitely 0.30.2.

So I'll close this issue. Thank you for your patience.

**Comment by blowmage** [comment]
Hi @hellupline and @vnbrs, we recently released version 0.32.0, with additional changes that should benefit long running subscribers. It would really help if you could try it out. We hope this is the last release before the gem goes 1.0.
